piqe_ocp_lib/api/resources/ocp_limit_ranges.py

```diff
@@ -0,0 +1,160 @@
+from piqe_ocp_lib.api.resources.ocp_base import OcpBase
+from kubernetes.client.rest import ApiException
+import logging
+from piqe_ocp_lib import __loggername__
+
+logger = logging.getLogger(__loggername__)
+
+
+class OcpLimitRanges(OcpBase):
+    """
+    OcpLimitRange Class extends OcpBase and encapsulates all methods
+    related to openshift limit ranges.
+    :param kube_config_file: A kubernetes config file.
+    :return: None
+    """
+    def __init__(self, kube_config_file=None):
+        super(OcpLimitRanges, self).__init__(kube_config_file=kube_config_file)
+        self.api_version = 'v1'
+        self.kind = 'LimitRange'
+        self.ocp_limit_range = self.dyn_client.resources.get(api_version=self.api_version, kind=self.kind)
+
+    def get_limit_ranges(self, namespace):
+        """
+        Get all LimitRanges from specific namespace
+        :param namespace: (str) name of the namespace
+        :return: LimitRanges response on Success OR None on Failure
+        """
+        api_response = None
+        try:
+            api_response = self.ocp_limit_range.get(namespace=namespace)
+        except ApiException as e:
+            logger.error(f"Exception while getting LimitRanges : {e}\n")
+
+        return api_response
+
+    def get_limit_ranges_names(self, namespace):
+        """
+        Get names of all LimitRanges from specific namespace
+        :param namespace: (str) name of the namespace
+        :return: List of LimitRanges names on Success OR Empty list on Failure
+        """
+        list_of_lr_names = list()
+        lr_response = self.get_limit_ranges(namespace=namespace)
+
+        if lr_response:
+            for lr in lr_response.items:
+                list_of_lr_names.append(lr.metadata.name)
+        else:
+            logger.warning(f"There are no LimitRanges in {namespace} namespace")
+
+        return list_of_lr_names
+
+    def delete_limit_ranges(self, namespace):
+        """
+        Delete the collection of LimitRange from specified namespace
+        :param namespace: (str) name of namespace where LimitRange was created
+        :return: Delete response on success OR None on Failure
+
+        ResourceInstance[Status]:
+          apiVersion: v1
+          details:
+            kind: limitrange
+            name: test
+            uid: aa1a8359-42b0-45f9-b44b-cd0ecbff6ef8
+          kind: Status
+          metadata: {}
+          status: Success
+
+        """
+        api_response = None
+        try:
+            api_response = self.ocp_limit_range.delete(namespace)
+        except ApiException as e:
+            logger.exception(f"Exception while deleting LimitRanges: {e}\n")
+        return api_response
+
+    def create_a_limit_range(self, limit_ranges_body):
```

Comment:
It might be worth looking into using the resource object models provided by the kubernetes client to obtain the body of this object. Then this would open the door to break this into more fine granular APIs that are more targeted, maybe by type of limit.
https://docs.openshift.com/container-platform/4.6/rest_api/objects/index.html#limitrangeitem-core-v1
https://github.com/kubernetes-client/python/blob/818e5ba0ba8cfc6d866195de706f58b7431110c1/kubernetes/client/models/v1_limit_range_item.py
It would go something like this:
Python 3.7.6 (default, Dec 19 2019, 22:50:01) 
[GCC 9.2.1 20190827 (Red Hat 9.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from kubernetes.client.models.v1_limit_range import V1LimitRange 
>>> lm = V1LimitRange()
>>> dir(lm)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_api_version', '_kind', '_metadata', '_spec', 'api_version', 'attribute_map', 'discriminator', 'kind', 'metadata', 'openapi_types', 'spec', 'to_dict', 'to_str']
>>> 
Ping me to discuss in more detail.

Reply:
Yup, I had similar thoughts about utilizing kubernetes/openshift models in our libraries. Do we want to abstract that away from the consumer so that they don't have to import the models themselves to pass in the model object as a parameter? They would just pass us the function keys as parameters and the function would generate the model and populate it for submission?

For this particular method signature I was following the status set for by the OcpConfigMaps resource which for their create function takes a dictionary. That means if we do decide to do this. Then we should probably do this for the other resource APIs, where feasible, so we are consistent across the board.

Reply:
I agree, the only aspect that makes me hesitant is that this would only be feasible with native kubernetes resources. I believe that ever since the switch was made to the dynamic client, Openshift specific models were no longer made available. I do like the direction you took with this and we should make a collective decision.